src/TableFunctions/TableFunctionObjectStorage.cpp

```diff
@@ -52,12 +54,50 @@ ObjectStoragePtr TableFunctionObjectStorage<Definition, Configuration, is_data_l
 }
 
 template <typename Definition, typename Configuration, bool is_data_lake>
-StorageObjectStorageConfigurationPtr TableFunctionObjectStorage<Definition, Configuration, is_data_lake>::getConfiguration() const
+StorageObjectStorageConfigurationPtr TableFunctionObjectStorage<Definition, Configuration, is_data_lake>::getConfiguration(ContextPtr context) const
 {
     if (!configuration)
     {
         if constexpr (is_data_lake)
-            configuration = std::make_shared<Configuration>(settings);
+        {
+            if (context->getSettingsRef()[Setting::datalake_disk_name].changed)
+            {
+                if (Definition::name != "iceberg" && Definition::name != "icebergCluster" && Definition::name != "deltaLake" && Definition::name != "deltaLakeCluster")
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Iceberg with disk configuration is supported only for default iceberg or icebergCluster");
+                auto disk = context->getDisk(context->getSettingsRef()[Setting::datalake_disk_name].value);
+                switch (disk->getObjectStorage()->getType())
+                {
+#if USE_AWS_S3 && USE_AVRO
+                case ObjectStorageType::S3:
+                    if (std::string_view(Definition::name).starts_with("iceberg"))
+                        configuration = std::make_shared<StorageS3IcebergConfiguration>(settings);
+                    else
+                        configuration = std::make_shared<StorageS3DeltaLakeConfiguration>(settings);
+                    break;
+#endif
+#if USE_AZURE_BLOB_STORAGE && USE_AVRO
+                case ObjectStorageType::Azure:
+                    if (std::string_view(Definition::name).starts_with("iceberg"))
+                        configuration = std::make_shared<StorageAzureIcebergConfiguration>(settings);
```

Comment:
If I was implementing this, I'd allow any type of storage, e.g. `iceberg`, `icebergS3`, `icebergAzure` to be allowed to be created with setting `datalake_disk_name` and just validate that function type conforms with disk storage type. I think this will be much better like I described and less confusing for users.